### PR TITLE
BlackBody model

### LINKED ISSUE #9066: Rationalize blackbody models/functions
There are three ways to calculate the emission from a blackbody in astropy.modeling - BlackBody1D, blackbody_lambda, and blackbody_nu.  The first is a full fittable 1D model and the latter two are functions.  The same answer is obtained with BlackBody1D and blackbody_nu (as stated in the docs).  Having 3 ways seems at least 1 too many.

It would be more consistent if any and all blackbody models in modeling were full fittable 1D models.  Mainly as all the rest of the models in modeling are modeling.models, not functions.

Probably it would be good to keep the two functions, but should deprecate them.

Additional minor comment.  BlackBody1D has two parameters - temperature and bolometric luminosity.  The 2nd parameter is more specific to (likely) stellar astrophysics - other areas of astronomy use other scalings usually a simple multiplicative amplitude (would allow other units).  It would be good to support a more generic way of supplying the amplitude (instead of bolometric or in addition).

### PR DESCRIPTION
**Pull request by karllark**
BlackBody model that is a straightforward implementation of Planck's Law.   This implementation is more flexible than the existing BlackBody1D model.

Added surface brightness equivalencies to the spectral_density equivalencies.  Useful for BlackBody and in general.

Docs and tests are included in this PR.

The BlackBody is the 1st model located in physical_models.py.  Other physics motivated models should go in this file (distinguished from functional_models.py that are those that are more mathematically motivated).  A later PR can/could move some models from functional to physical as appropriate.

This new function replaces the existing blackbody_lambda and blackbody_nu functions and BlackBody1D model.  See #9066 for discussion of why this is desirable.  These existing functions have been deprecated as part of this PR.  Documentation showing how to reproduce the results from all three with BlackBody has been included.

Closes #9066

### CODE REVIEW

File: astropy/modeling/blackbody.py [hunk 1]
```diff
@@ -1,109 +1,6 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 """
 Model and functions related to blackbody radiation.
-
-.. _blackbody-planck-law:
```
Comment (pllim) [inline astropy/modeling/blackbody.py]: Please move this ref to the new section you added. Thanks!

**Comment by bsipocz** [comment]
@pllim - please remove the label once you're happy with it downstream.

todo for the bot: stop the merge button to appear while a PR has  the `needs-downstream-testing` label.

File: CHANGES.rst [hunk 2]
```diff
@@ -69,14 +69,15 @@ astropy.modeling
 - Significant reorganization of the documentation. [#9078, #9171]
 
 - Add ``Tabular1D.inverse`` [#9083]
-- Significant reorganization of the documentation. [#9078, #9171]
```
Comment (pllim) [inline CHANGES.rst]: Removing duplicate change log entry should be a separate PR. Unless this was not intentional?

File: CHANGES.rst [hunk 3]
```diff
@@ -69,14 +69,15 @@ astropy.modeling
...
 
 - ``Model.rename`` was changed to add the ability to rename ``Model.inputs``
   and ``Model.outputs``. [#9220]
 
 - New function ``fix_inputs`` to generate new models from others by fixing
   specific inputs variable values to constants. [#9135]
 
+- Add BlackBody model.  [#9282]
```
Comment (pllim) [inline CHANGES.rst]:
```suggestion
- Add ``BlackBody`` model.
```

File: CHANGES.rst [hunk 4]
```diff
@@ -69,14 +69,15 @@ astropy.modeling
...
+  Deprecate BlackBody1D model and blackbody_nu and blackbody_lambda functions.
```
Comment (pllim) [inline CHANGES.rst]:
```suggestion
  Deprecate ``BlackBody1D`` model and ``blackbody_nu`` and ``blackbody_lambda`` functions. [#9282]
```

File: CHANGES.rst [hunk 5]
```diff
@@ -84,6 +84,8 @@ astropy.modeling
   ``inputs`` and ``outputs`` are generated. ``Model.inputs`` and ``Model.outputs``
   are now settable which allows renaming them on per user case. [#9298]
 
+- Add ``BlackBody`` model.  [#9282]
```
Comment (pllim) [inline CHANGES.rst]: p.s. The usual convention is to show PR number at the end of the last line of change log entry. See other entries as an example.

File: docs/modeling/physical_models.rst [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    bb_nu = bb(wavelengths)
+    bb_lambda = (bb_nu * u.sr).to(
+        u.erg / (u.cm * u.cm * u.Angstrom * u.s),
+        equivalencies=u.spectral_density(wavelengths),
+    ) / u.sr
+
+The ``bolometric_flux()`` member function gives the bolometric flux using
+:math:`\sigma T^4/\pi` where :math:`\sigma` is the Stefan-Boltzmann constant.
+
+.. note::
```
Comment (pllim) [inline docs/modeling/physical_models.rst]: I think this note should be moved to the top, maybe right after the first heading.

File: docs/modeling/physical_models.rst [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+    Prior to v4.0, the BlackBody1D and the functions blackbody_nu and blackbody_lambda
```
Comment (pllim) [inline docs/modeling/physical_models.rst]:
```suggestion
    Prior to v4.0, ``BlackBody1D``, ``blackbody_nu``, and ``blackbody_lambda``
```

File: docs/modeling/physical_models.rst [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    were provided.  BlackBody1D was a more limited blackbody model that was
```
Comment (pllim) [inline docs/modeling/physical_models.rst]:
```suggestion
    were provided.  ``BlackBody1D`` was a more limited blackbody model that was
```

File: docs/modeling/blackbody_deprecated.rst [hunk 9]
```diff
@@ -0,0 +1,129 @@
+:orphan:
+
+Blackbody (deprecated capabilities)
+===================================
+
+blackbody functions (deprecated)
```
Comment (pllim) [inline docs/modeling/blackbody_deprecated.rst]:
```suggestion
Blackbody functions (deprecated)
```

File: docs/modeling/physical_models.rst [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+:math:`k` is the Boltzmann constant.
+
+The two parameters of the model the unitless scaling factor ``scale`` and
+the absolute temperature ``temperature``.  The ``scale`` factor is unitless
+as the Plank function has natural units of spectral radiance, specifically
+ergs/(cm^2 Hz s sr).  The
+temperature can be passed as a Quantity with any temperature unit.  If temperature
+is passed without a unit, it is assumed to be in Kelvin units.
+
+An example plot for a blackbody with a temperature of 10000 K and a scale of 1 is
```
[the remark this answers is not in the document]
Comment (karllark) [reply docs/modeling/physical_models.rst]: Added a sentence with this info.

File: docs/modeling/physical_models.rst [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    ax.set_xlabel(r"$\lambda$ [{}]".format(wavelengths.unit))
+    ax.set_ylabel(r"$Flux$ [{}]".format(bb_result.unit))
+
+    plt.tight_layout()
+    plt.show()
+
+The Planck function can be needed as :math:`B_\lambda` in units of
+ergs/(cm^2 s A sr).  This can be easily accomplished with a
+unit conversion as illustrated below.  Note that this uses the ``spectral_density``
+equivalencies requiring the steradian units to be first removed, then reapplied
```
[the remark this answers is not in the document]
Comment (karllark) [reply docs/modeling/physical_models.rst]: Yep.

File: docs/modeling/physical_models.rst (hunk 6 above)
Comment (karllark) [reply docs/modeling/physical_models.rst]: I prefer to be less prominent.  Better to motivate others to not use it at all!

File: docs/modeling/physical_models.rst [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+where :math:`\nu` is the frequency, :math:`T` is the temperature,
+:math:`A` is the scaling factor,
+:math:`h` is the Plank constant, :math:`c` is the speed of light, and
+:math:`k` is the Boltzmann constant.
+
+The two parameters of the model the unitless scaling factor ``scale`` and
+the absolute temperature ``temperature``.  The ``scale`` factor is unitless
+as the Plank function has natural units of spectral radiance, specifically
+ergs/(cm^2 Hz s sr).  The
```
[the remark this answers is not in the document]
Comment (karllark) [reply docs/modeling/physical_models.rst]: Nope.  Does not look better at all.

File: astropy/units/equivalencies.py [hunk 13]
```diff
@@ -263,6 +267,8 @@ def iconverter_phot_f_nu_to_f_la(x):
         (phot_L_la, phot_L_nu, converter_phot_L_la_phot_L_nu, iconverter_phot_L_la_phot_L_nu),
         (phot_L_nu, L_nu, converter_phot_L_nu_to_L_nu, iconverter_phot_L_nu_to_L_nu),
         (phot_L_nu, L_la, converter_phot_L_nu_to_L_la, iconverter_phot_L_nu_to_L_la),
+        # surface brightness
+        (S_la, S_nu, converter, iconverter),
```
Comment (pllim) [inline astropy/units/equivalencies.py]: cc @mhvk for the new unit equivalency.

File: astropy/modeling/tests/test_physical_models.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    assert len(w) == 1
+    assert "invalid" in w[0].message.args[0]
+
+
+def test_blackbody_array_temperature():
+    """Regression test to make sure that the temperature can be an array."""
+    multibb = BlackBody([100, 200, 300] * u.K)
+    flux = multibb(1.2 * u.mm)
+    np.testing.assert_allclose(
+        flux.value, [1.804908e-12, 3.721328e-12, 5.638513e-12], rtol=1e-5
```
Comment (pllim) [inline astropy/modeling/tests/test_physical_models.py]: Should check the unit as well.

File: astropy/modeling/tests/test_physical_models.py [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    )
+
+    flux = multibb([2, 4, 6] * u.mm)
+    np.testing.assert_allclose(
+        flux.value, [6.657915e-13, 3.420677e-13, 2.291897e-13], rtol=1e-5
+    )
+
+    multibb = BlackBody(np.ones(4) * u.K)
+    flux = multibb(np.ones((3, 4)) * u.mm)
+    assert flux.shape == (3, 4)
```
Comment (pllim) [inline astropy/modeling/tests/test_physical_models.py]: Why not also check for the values and units instead just the shape?

**Comment by pllim** [comment]
As discussed offline, the overflow test needs to be adapted for the refactored module as well.

Also, I wonder if this test needs to be ported over as well. It was originally adapted from #1480.

https://github.com/astropy/astropy/blob/09b7796bbd4eadc793bb48fc5c8554bb5bd4af0d/astropy/modeling/tests/test_blackbody.py#L56-L73

**Comment by pllim** [comment]
Note for the future: This refactoring gets rid of the ability to fit by bolometric flux (#4855) and also the `lambda_max` property. I am neutral on this. Just mainly a note for my future self.

File: astropy/units/equivalencies.py (hunk 13 above)
Comment (mhvk) [reply astropy/units/equivalencies.py]: The addition itself is fine, but I think it would make sense to do also `phot_S_*`, `nu_S_nu`, etc., as for flux and luminosity.

Comment (pllim) [reply astropy/units/equivalencies.py]: I agree with @mhvk as I failed to convert from `BlackBody` flux output to PHOTLAM/sr. I get `UnitConversionError: 'erg / (cm2 Hz s sr)' and 'ph / (Angstrom cm2 s sr)' are not convertible`.

I have to work around it by manually getting rid of `sr`.

**Comment by karllark** [comment]
@pllim : review (all of it) much appreciated.  I'm doing a bit of work on the PR right now.  You may be happy to learn that our in room conversations did convince me to put in lambda_max (and nu_max) as properties.  Sometimes, I just need time.  ;-)

File: astropy/modeling/tests/test_physical_models.py (hunk 15 above)
Comment (karllark) [reply astropy/modeling/tests/test_physical_models.py]: The values and units returned by BlackBody are checked in earlier tests.  This test existed for BlackBody1D and so I ported it over for BlackBody.  I'm assuming the goal of this test is to test if BlackBody can accept arrays for temperature.

File: astropy/units/equivalencies.py (hunk 13 above)
Comment (karllark) [reply astropy/units/equivalencies.py]: I have added in all the equivalent surface brightness conversions to the existing flux and luminosity.  I have not added tests for all of them as they use the same converters so that flux/luminosity versions of the tests should be good (I think).  If tests are needed, I can add them through extensive copy and paste from existing tests (or maybe there is a better way).

**Comment by nden** [comment]
@astrofrog Since you wrote the original one - do you have time to review this?

File: astropy/modeling/physical_models.py [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+    @property
+    def lambda_max(self):
+        """Peak wavelength when the curve is expressed as power density."""
+        return const.b_wien / self.temperature
+
+    @property
+    def nu_max(self):
+        """Peak frequency when the curve is expressed as power density."""
+        return 2.8214391 * const.k_B * self.temperature / const.h
```
Comment (pllim) [inline astropy/modeling/physical_models.py]: Interesting. I thought you said exact calculation of this was not possible. What is 2.8214391? Is there some literature reference that is tied to this formula?

Comment (karllark) [reply astropy/modeling/physical_models.py]: Yep.  I put it in the docs.  It is https://en.wikipedia.org/wiki/Wien%27s_displacement_law

File: docs/modeling/physical_models.rst [hunk 17]
```diff
@@ -66,9 +66,14 @@ default units returned by :class:`~astropy.modeling.physical_models.BlackBody`.
     plt.tight_layout()
     plt.show()
 
-The ``bolometric_flux()`` member function gives the bolometric flux using
+The ``bolometric_flux`` member function gives the bolometric flux using
```
Comment (pllim) [inline docs/modeling/physical_models.rst]: Why not use Sphinx API linking like this?
```
:meth:`~astropy.modeling.physical_models.BlackBody.bolometric_flux`
```
Same comment applies to other functions with API doc that is not linked.

File: astropy/units/equivalencies.py [hunk 18]
```diff
@@ -267,8 +279,24 @@ def iconverter_phot_f_nu_to_f_la(x):
...
+        # surface brightness (flux equiv)
         (S_la, S_nu, converter, iconverter),
+        (S_nu, nu_S_nu, converter_f_nu_to_nu_f_nu, iconverter_f_nu_to_nu_f_nu),
+        (S_la, la_S_la, converter_f_la_to_la_f_la, iconverter_f_la_to_la_f_la),
+        (phot_S_la, S_la, converter_phot_f_la_to_f_la, iconverter_phot_f_la_to_f_la),
+        (phot_S_la, S_nu, converter_phot_f_la_to_f_nu, iconverter_phot_f_la_to_f_nu),
+        (phot_S_la, phot_S_nu, converter_phot_f_la_phot_f_nu, iconverter_phot_f_la_phot_f_nu),
+        (phot_S_nu, S_nu, converter_phot_f_nu_to_f_nu, iconverter_phot_f_nu_to_f_nu),
+        (phot_S_nu, S_la, converter_phot_f_nu_to_f_la, iconverter_phot_f_nu_to_f_la),
+        # surface brightness (luminosity equiv)
```
Comment (pllim) [inline astropy/units/equivalencies.py]: Are all the new converters tested?

Comment (karllark) [reply astropy/units/equivalencies.py]: No.  As they are the same converters as for the flux density units, they are tested there.  Does not seem like they need to be tested a 2nd time as the only change is to add a per steradian units to both sides of the test.

File: astropy/modeling/physical_models.py [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+        from astropy import units as u
+        from astropy.visualization import quantity_support
+
+        bb = BlackBody(temperature=5778*u.K)
+        wav = np.arange(1000, 110000) * u.AA
+        flux = bb(wav)
+
+        with quantity_support():
+            plt.figure()
+            plt.semilogx(wav, flux)
```
Comment (pllim) [inline astropy/modeling/physical_models.py]: Now that you added method for lambda max, maybe plot it as well? :thinking:

File: docs/modeling/physical_models.rst (hunk 17 above)
Comment (karllark) [reply docs/modeling/physical_models.rst]: Will do.

File: astropy/units/equivalencies.py (hunk 18 above)
Comment (pllim) [reply astropy/units/equivalencies.py]: I'll defer to @mhvk .

**Comment by karllark** [comment]
@pllim : Yep.  I'm fixing this to going back to the behavior that initializing BlackBody requires temperature as a quantity, but calling the evaluate function does not.  This solves the problem of not having units on temperature.  This does remove the capability of initializing the BlackBody with an assumed unit on temperature, but I don't see how to support this *and* have the right units on temperature for the member functions (properties).  This is what BlackBody1D did so I think this is reasonable for BlackBody to do the same.

**Comment by pllim** [comment]
@karllark , yes, that is reasonable. Looks like I finally have to pay my technical debt over at `synphot` side... :see_no_evil:

File: astropy/modeling/physical_models.py (hunk 19 above)
Comment (karllark) [reply astropy/modeling/physical_models.py]: Done.

File: astropy/modeling/physical_models.py [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+    def input_units(self):
+        # The input units are those of the 'x' value, which should always be
+        # Hz. Because we do this, and because input_units_allow_dimensionless
+        # is set to True, dimensionless values are assumed to be in Hz.
+        return {"x": u.Hz}
+
+    def _parameter_units_for_data_units(self, inputs_unit, outputs_unit):
+        return OrderedDict([("temperature", u.K)])
+
+    # does not work as need to pass freq/wave for unit equivalencies
```
Comment (pllim) [inline astropy/modeling/physical_models.py]: Did you ever figure this one out? If not, please remove the commented code.

File: astropy/modeling/physical_models.py [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+
+        Raises
+        ------
+        ValueError
+            Invalid temperature.
+
+        ZeroDivisionError
+            Wavelength is zero (when converting to frequency).
+        """
+        if not isinstance(temperature, u.Quantity):
```
Comment (pllim) [inline astropy/modeling/physical_models.py]: Why not attach this to `input_units`?

Comment (karllark) [reply astropy/modeling/physical_models.py]: Need to know if temperature is passed with units or not in the code so that the results can be passed w/o units if temp pasted w/o units (supports the fitting).

File: astropy/modeling/physical_models.py (hunk 20 above)
Comment (karllark) [reply astropy/modeling/physical_models.py]: Can't do what is needed as equivalencies need to be used and this is not possible with the current return_units code.

Comment (karllark) [reply astropy/modeling/physical_models.py]: Commented code removed.

File: astropy/modeling/physical_models.py [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+            plt.show()
+    """
+
+    # We parametrize this model with a temperature and a scale.
+    temperature = Parameter(default=5000.0, min=0, unit=u.K)
+    scale = Parameter(default=1.0, min=0)
+
+    # We allow values without units to be passed when evaluating the model, and
+    # in this case the input x values are assumed to be frequencies in Hz.
+    _input_units_allow_dimensionless = True
```
Comment (pllim) [inline astropy/modeling/physical_models.py]: Is there supposed to be a leading underscore? See https://docs.astropy.org/en/latest/modeling/add-units.html#input-units-allow-dimensionless

Comment (nden) [reply astropy/modeling/physical_models.py]: Yes, the one you are pointing to is a property. There are still loose ends with units that we won't have time to fix for 4.0

File: astropy/modeling/physical_models.py [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+
+    @property
+    def input_units(self):
+        # The input units are those of the 'x' value, which should always be
+        # Hz. Because we do this, and because input_units_allow_dimensionless
+        # is set to True, dimensionless values are assumed to be in Hz.
+        return {"x": u.Hz}
+
+    def _parameter_units_for_data_units(self, inputs_unit, outputs_unit):
+        return OrderedDict([("temperature", u.K)])
```
Comment (pllim) [inline astropy/modeling/physical_models.py]: Is `OrderedDict` really necessary with one element?

File: astropy/modeling/physical_models.py [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+
+    @property
+    def bolometric_flux(self):
+        """Bolometric flux."""
+        return const.sigma_sb * self.temperature ** 4 / np.pi
```
Comment (pllim) [inline astropy/modeling/physical_models.py]: Hmm, so the bolometric flux does not take the fitted scale into account? Talking to you the other week, I was under the impression that this would be derived from both temperature and scale.

File: docs/modeling/physical_models.rst [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+<https://en.wikipedia.org/wiki/Wien%27s_displacement_law>`_.
+
+.. note::
+
+    Prior to v4.0, the ``BlackBody1D`` and the functions ``blackbody_nu`` and ``blackbody_lambda``
+    were provided.  ``BlackBody1D`` was a more limited blackbody model that was
+    specific to integrated fluxes from sources.  See :doc:`blackbody_deprecated`
+    and `astropy issue #9066 <https://github.com/astropy/astropy/issues/9066>`_ for details.
+    The capabilities of all three
+    can be obtained with :class:`~astropy.modeling.physical_models.BlackBody`.
```
Comment (astrofrog) [inline docs/modeling/physical_models.rst]: Is there any way to demonstrate how? (in the deprecated doc page)

File: astropy/modeling/physical_models.py (hunk 23 above)
Comment (karllark) [reply astropy/modeling/physical_models.py]: Nope.  Removed.

File: astropy/modeling/physical_models.py (hunk 24 above)
Comment (karllark) [reply astropy/modeling/physical_models.py]: Good catch!  Scale definitely should be used.  Added it.

File: docs/modeling/physical_models.rst (hunk 25 above)
Comment (karllark) [reply docs/modeling/physical_models.rst]: Yes.  Should have done this when @pllim asked for it clearly.  Done now.

**Comment by karllark** [comment]
I believe I've addressed all the comments.  Hopefully we are converging.  :-)

**Comment by karllark** [comment]
I don't understand the issue with building the docs.  They work fine when I create them on my laptop.  Maybe a version issue?  And I can't figure out the units issue from the error log.

**Comment by pllim** [comment]
Any situation where your example would generate something in `1 / (Hz sr)`?
```
astropy.units.core.UnitConversionError: '1 / (Hz sr)' and 'erg / (cm2 Hz s)' (spectral flux density) are not convertible
```

Did you clean your doc build before rebuilding? Sometimes you could be skipping that example because an older (successful) cache already exists.

Anyway, looks like you need a rebase, as there is a conflict in change log.

**Comment by karllark** [comment]
> Anyway, looks like you need a rebase, as there is a conflict in change log.

Just did this earlier today for the same reason.  Then another modeling PR was merged making a conflict reappear.  I'll wait a little bit and see if there are any other changes needed and then rebase.  :-)

**Comment by nden** [comment]
@karllark Sorry for the multiple conflicts caused. That's it from me today. Please squash your commits when you merge. This looks good and as far as I can tell it's ready to merge but I'm leaving the honor to @pllim .

**Comment by karllark** [comment]
> Any situation where your example would generate something in `1 / (Hz sr)`?
> 
> ```
> astropy.units.core.UnitConversionError: '1 / (Hz sr)' and 'erg / (cm2 Hz s)' (spectral flux density) are not convertible
> ```
> 
> Did you clean your doc build before rebuilding? Sometimes you could be skipping that example because an older (successful) cache already exists.

Cleaned, rebuild fine.  Checked plotting code manually by running it in file.  No idea where this error is from.  I'm going to rebase, squash down to a reasonable set of commits, and then push.  Maybe the error will go away (unlikely).  If it doesn't, I don't know what to try next.

**Comment by bsipocz** [comment]
I haven't yet tested the failing stuff with this branch, but couldn't not notice that you add docs for the deprecated functionality? Is that necessary? 

I mean we shouldn't really advertize and thus document now deprecated stuff when there is another suggested way to get to the same result.

**Comment by bsipocz** [comment]
OK, my local docs build is not working (due to other reasons), but the example plots works as expected on python3.7 and numpy dev.

File: astropy/modeling/physical_models.py (hunk 19 above)
Comment (bsipocz) [reply astropy/modeling/physical_models.py]: I'm not sure how useful it is to have plots in the API  docs. How does user normally interact with this? From the terminal/notebook they won't be able to see the plots, do they?

**Comment by bsipocz** [comment]
Based on the tracebacks and generated artefacts, I'm fairly certain that the failure is triggered by the first example, specifically this rendering of it:
https://47429-2081289-gh.circle-artifacts.com/0/home/circleci/project/docs/_build/html/modeling/blackbody_deprecated-1.py

I cannot make it fail locally though with any of the version combinations I tried so far.

**Comment by pllim** [comment]
Ah, I get it! Somewhere else in the documentation build (remember that namespace is shared for all the documentation pages), this happened:
```python
>>> from astropy.visualization import quantity_support
>>> quantity_support()  
```

As a result, it tried to convert `bb(wavelengths)/bb.bolometric_flux` (this has unit of `1 / (Hz sr)`) to the unit of `bb1d(wavelengths)` (this has unit of `erg / (cm2 Hz s)`) and failed. So, it emits a warning with the traceback but plots the stuff anyway. In our CI, we convert all Sphinx warnings into failures. And there you have it.

That said, if `bb(wavelengths)/bb.bolometric_flux` and `bb1d(wavelengths)` are supposed to be equivalent, how do we explain that when they don't have the same units? @karllark , I know you tried to explain it to me but it slipped my grasp again. 😬

**Comment by bsipocz** [comment]
Oh, great. Once we find some of those infinite free time we might want to fix that docs pages are independent, etc.

However, my question about advertizing the deprecated behaviour still stands. Shouldn't it be removed fully?

File: docs/modeling/blackbody_deprecated.rst [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+    spectrum_nu = blackbody_nu(wavelengths, temperature)
+
+    # provides the result in ergs/(cm^2 A s sr)
+    spectrum_lam = blackbody_lambda(wavelengths, temperature)
+
+    # default results from BlackBody same units as blackbody_nu
+    bb_nu = BlackBody(temperature)
+
+    # use scale to change results units for Blackbody to match blackbody_lambda
+    bb_lam = BlackBody(temperature, scale=1.0 * u.erg / (u.cm ** 2 * u.AA * u.s * u.sr))
```
Comment (pllim) [inline docs/modeling/blackbody_deprecated.rst]: Interesting... I wouldn't have thought of using `scale` in that way. I would have done something more involved, like:
```python
 bb_nu(wavelengths).to(u.erg / (u.cm ** 2 * u.AA * u.s * u.sr), u.spectral_density(wavelengths))
```
So, thanks for including this example!

Comment (karllark) [reply docs/modeling/blackbody_deprecated.rst]: Using scale this way make it "easier" for the user *and* (after my latest commit) also means that the scale units can be used to set the output units even during fitting.  This can't be done using the return_units functionality of modeling as spectral_density equivalencies is needed and this isn't support by return_units at this point.

**Comment by karllark** [comment]
Latest commit should fix the html docs bug (if I understood it correctly). Minor text cleanup as well.  

This commit also means that the results are returned in scale units even during fitting (been trying to figure out how to make this happen for sometime and a solution finally dropped into my brain).  

Hopefully this is the last commit needed (fingers-crossed).

File: docs/modeling/blackbody_deprecated.rst (hunk 26 above)
Comment (karllark) [reply docs/modeling/blackbody_deprecated.rst]: What you would have done works as well of course.

**Comment by nden** [comment]
Everything looks in order and approved, so I'm merging it. Thanks @karllark !
